gguf-py/scripts/gguf-dump.py

```diff
@@ -101,25 +101,200 @@ def dump_metadata_json(reader: GGUFReader, args: argparse.Namespace) -> None:
+                    for element_pos in range(render_element):
+                        value += repr(str(bytes(field.parts[-1 - element_pos]), encoding='utf-8')[:5]) + (", " if total_elements > 1 else "")
+                elif curr_type in reader.gguf_scalar_to_np:
+                    render_element = min(7, total_elements)
+                    for element_pos in range(render_element):
+                        value += str(field.parts[-1 - element_pos][0]) + (", " if total_elements > 1 else "")
+                value = f'[ {value}{" ..." if total_elements > 1 else ""} ]'
+        markdown_content += f'| {n:3} | {pretty_type:10} | {total_elements:8} | {field.name:38} | {value:<78} |\n'
+
+    markdown_content += "\n"
+
+    if not args.no_tensors:
+        # Group tensors by their prefix and maintain order
+        tensor_prefix_order = []
+        tensor_name_to_key = {}
+        tensor_groups = {}
+        total_elements = sum(tensor.n_elements for tensor in reader.tensors)
+
+        for key, tensor in enumerate(reader.tensors):
+            tensor_components = tensor.name.split('.')
+            tensor_prefix = tensor_components[0]
+
+            if tensor_prefix == 'blk':
+                tensor_prefix = f"{tensor_components[0]}.{tensor_components[1]}"
+
+            if tensor_prefix not in tensor_groups:
+                tensor_groups[tensor_prefix] = []
+                tensor_prefix_order.append(tensor_prefix)
+
+            tensor_name_to_key[tensor.name] = key
+            tensor_groups[tensor_prefix].append(tensor)
+
+        # Tensors Mapping Dump
+        markdown_content += f'## Tensors Overview {element_count_rounded_notation(total_elements)} Elements\n'
+        markdown_content += f'Total number of elements in all tensors: {total_elements} Elements\n'
+        markdown_content += '\n'
+
+        for group in tensor_prefix_order:
+            tensors = tensor_groups[group]
+            group_elements = sum(tensor.n_elements for tensor in tensors)
+            markdown_content += f"- [{translate_tensor_name(group)} Tensor Group - {element_count_rounded_notation(group_elements)} Elements](#{group.replace('.', '_')})\n"
+
+        markdown_content += "\n"
+
+        for group in tensor_prefix_order:
+            tensors = tensor_groups[group]
+            group_elements = sum(tensor.n_elements for tensor in tensors)
+            group_percentage = group_elements / total_elements * 100
+            markdown_content += f"### {translate_tensor_name(group)} Tensor Group : {element_count_rounded_notation(group_elements)} Elements <a name=\"{group.replace('.', '_')}\"></a>\n"
+            markdown_content += "| T_ID | Tensor Layer Name         | Human Friendly Tensor Layer Name                   | Elements       | Shape                           | Type |\n"
+            markdown_content += "|------|---------------------------|----------------------------------------------------|----------------|---------------------------------|------|\n"
```

Comment:
Should the column widths be dynamic depending on what will be printed? Otherwise, I get some quite ugly formatting, especially for `ffn_gate_inp`:

```markdown
### Block 7 Tensor Group : ~108M Elements {#blk_7}
| T_ID | Tensor Layer Name         | Human Friendly Tensor Layer Name                   | Elements       | Shape                           | Type |
|------|---------------------------|----------------------------------------------------|----------------|---------------------------------|------|
|  136 | blk.7.ffn_down_exps.weight | Block 7 Ffn_Down_Exps (W)                          | (~34M) 33554432 | [4096  x 1024  x   8   x   1  ] | Q8_0 |
|  137 | blk.7.ffn_gate_exps.weight | Block 7 Ffn_Gate_Exps (W)                          | (~34M) 33554432 | [1024  x 4096  x   8   x   1  ] | Q8_0 |
|  138 | blk.7.ffn_up_exps.weight  | Block 7 Ffn_Up_Exps (W)                            | (~34M) 33554432 | [1024  x 4096  x   8   x   1  ] | Q8_0 |
|  139 | blk.7.ffn_gate_inp.weight | Block 7 Expert-Routing Layer For The Feed-Forward Network In Mixture Of Expert Models (W) | ( ~8K)    8192 | [1024  x   8   x   1   x   1  ] | F32  |
|  140 | blk.7.attn_norm.weight    | Block 7 Attention Normalization (W)                | ( ~1K)    1024 | [1024  x   1   x   1   x   1  ] | F32  |
|  141 | blk.7.ssm_a               | Block 7 State Space Model State Compression        | (~33K)   32768 | [ 16   x 2048  x   1   x   1  ] | F32  |
|  142 | blk.7.ssm_d               | Block 7 State Space Model Skip Connection          | ( ~2K)    2048 | [2048  x   1   x   1   x   1  ] | F32  |
|  143 | blk.7.ssm_b_norm.weight   | Block 7 Ssm_B_Norm (W)                             | (  16)      16 | [ 16   x   1   x   1   x   1  ] | F32  |
|  144 | blk.7.ssm_c_norm.weight   | Block 7 Ssm_C_Norm (W)                             | (  16)      16 | [ 16   x   1   x   1   x   1  ] | F32  |
|  145 | blk.7.ssm_conv1d.bias     | Block 7 State Space Model Rolling/Shift (B)        | ( ~2K)    2048 | [2048  x   1   x   1   x   1  ] | F32  |
|  146 | blk.7.ssm_conv1d.weight   | Block 7 State Space Model Rolling/Shift (W)        | ( ~8K)    8192 | [  4   x 2048  x   1   x   1  ] | F32  |
|  147 | blk.7.ssm_dt_norm.weight  | Block 7 Ssm_Dt_Norm (W)                            | ( 256)     256 | [ 256  x   1   x   1   x   1  ] | F32  |
|  148 | blk.7.ssm_dt.bias         | Block 7 State Space Model Time Step (B)            | ( ~2K)    2048 | [2048  x   1   x   1   x   1  ] | F32  |
|  149 | blk.7.ssm_dt.weight       | Block 7 State Space Model Time Step (W)            | (~524K)  524288 | [ 256  x 2048  x   1   x   1  ] | F32  |
|  150 | blk.7.ssm_in.weight       | Block 7 State Space Model Input Projections (W)    | ( ~4M) 4194304 | [1024  x 4096  x   1   x   1  ] | Q8_0 |
|  151 | blk.7.ssm_out.weight      | Block 7 State Space Model Output Projection (W)    | ( ~2M) 2097152 | [2048  x 1024  x   1   x   1  ] | Q8_0 |
|  152 | blk.7.ssm_x.weight        | Block 7 State Space Model Selective Parametrization (W) | (~590K)  589824 | [2048  x  288  x   1   x   1  ] | F32  |
|  153 | blk.7.ffn_norm.weight     | Block 7 Feed-Forward Network Normalization (W)     | ( ~1K)    1024 | [1024  x   1   x   1   x   1  ] | F32  |
```

Reply:
There are python modules, but gg tends to prefer avoiding adding dependencies.

Fortunately I created a python function specifically to solve this problem before. JSON to Markdown table formatting: https://stackoverflow.com/a/72983854/2850957